northstar/src/runtime/token.rs

```diff
@@ -0,0 +1,123 @@
+use blake3::Hash;
+use lazy_static::lazy_static;
+use rand_core::{OsRng, RngCore};
+use std::io::{Cursor, Write};
+use std::ops::Add;
+use std::time::{Duration, Instant};
+use thiserror::Error;
+
+// Tokens are valid for one minute
+const TOKEN_EXPIRED_THRESHOLD: Duration = Duration::from_secs(60);
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("Failed to write token")]
+    WriteToken,
+}
+
+lazy_static! {
+    static ref MAC_KEY: [u8; 32] = {
+        let mut key = [0u8; 32];
+        OsRng.fill_bytes(&mut key);
+        key
+    };
+    static ref FIRST_TOKEN_TIME: Instant = Instant::now();
+}
+
+pub type Timestamp = u64;
+
+pub struct Token {
+    time: Timestamp,
+    mac: blake3::Hash,
+}
+
+impl Token {
+    pub fn new(application: &str) -> Token {
+        let first = *FIRST_TOKEN_TIME; // Do not inline to ensure earliest initialization
+        let time = Instant::now().duration_since(first).as_secs();
+        let mac = Token::calculate_mac(&time, application);
+        Token { time, mac }
+    }
+
+    pub fn verify(&self, application: &str) -> bool {
+        // Check MAC
+        let mac = Token::calculate_mac(&self.time, application);
+        if self.mac != mac {
+            return false;
+        }
+
+        // Check expiration
+        let token_time = FIRST_TOKEN_TIME.add(Duration::from_secs(self.time));
+        if token_time < *FIRST_TOKEN_TIME {
+            return false; // Token is older than the very first one
+        }
+        let token_age = Instant::now().duration_since(token_time);
+        token_age <= TOKEN_EXPIRED_THRESHOLD
```

Comment:
nit: this would be easier to read with an `else` instead of the `return`.

Reply:
Line 54 panics if we do not bail on line 52. I am not sure how it can be rearranged.